modules/apps/headless/headless-builder/headless-builder-impl/src/main/java/com/liferay/headless/builder/internal/publisher/APIApplicationPublisherImpl.java

```diff
@@ -0,0 +1,113 @@
+import com.liferay.headless.builder.application.publisher.APIApplicationPublisher;
+import com.liferay.headless.builder.internal.jaxrs.application.HeadlessBuilderApplication;
+import com.liferay.portal.kernel.util.HashMapDictionaryBuilder;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import javax.ws.rs.core.Application;
+
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceRegistration;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+
+/**
+ * @author Luis Miguel Barcos
+ */
+@Component(service = APIApplicationPublisher.class)
+public class APIApplicationPublisherImpl implements APIApplicationPublisher {
+
+	@Override
+	public void publish(APIApplication apiApplication) {
+		_headlessBuilderApplicationServiceRegistrationMap.putIfAbsent(
+			apiApplication.getOSGiJaxRsName(),
+			_registerHeadlessBuilderApplication(apiApplication));
+	}
+
+	@Override
+	public void unpublish(APIApplication apiApplication) {
+		ServiceRegistration<Application> applicationServiceRegistration =
+			_headlessBuilderApplicationServiceRegistrationMap.remove(
+				apiApplication.getOSGiJaxRsName());
+
+		if (applicationServiceRegistration != null) {
+			_unregisterServiceRegistration(applicationServiceRegistration);
+		}
+	}
+
+	@Activate
+	protected void activate(BundleContext bundleContext) {
+		_bundleContext = bundleContext;
+	}
+
+	@Deactivate
+	protected void deactivate() {
+		_unpublishAll();
+	}
+
+	private ServiceRegistration<Application>
+		_registerHeadlessBuilderApplication(APIApplication apiApplication) {
+
+		return _bundleContext.registerService(
+			Application.class, new HeadlessBuilderApplication(),
+			HashMapDictionaryBuilder.<String, Object>put(
+				"companyId", apiApplication.getCompanyId()
+			).put(
+				"liferay.filter.disabled", true
+			).put(
+				"liferay.headless.builder.application", true
+			).put(
+				"liferay.jackson", false
+			).put(
+				"osgi.jaxrs.application.base", apiApplication.getBaseURL()
+			).put(
+				"osgi.jaxrs.extension.select",
+				"(osgi.jaxrs.name=Liferay.Vulcan)"
+			).put(
+				"osgi.jaxrs.name", apiApplication.getOSGiJaxRsName()
+			).build());
+	}
+
+	private void _unpublishAll() {
+		for (ServiceRegistration<Application> serviceRegistration :
+				_headlessBuilderApplicationServiceRegistrationMap.values()) {
+
+			_unregisterServiceRegistration(serviceRegistration);
+		}
+
+		_headlessBuilderApplicationServiceRegistrationMap.clear();
+	}
+
+	private void _unregisterServiceRegistration(
+		ServiceRegistration<?> serviceRegistration) {
+
+		if (serviceRegistration != null) {
+			serviceRegistration.unregister();
+		}
+	}
+
+	private static BundleContext _bundleContext;
+
+	private final Map<String, ServiceRegistration<Application>>
+		_headlessBuilderApplicationServiceRegistrationMap = new HashMap<>();
```

Comment:
My bet is that BC will change this name :money_with_wings:

Reply:
I would leave it as it is and let Brian decide de name